### ISSUE TITLE
[🐞]: Spawnradius Gamerule Causes Exponentially Long World Load Times

### ISSUE THREAD

**Issue by dandy-is-lion** [MehVahdJukaar/Supplementaries#1301]
### Before Continuing:

- [X] Check you are using the latest version of the mods and its dependencies
- [X] Always include a latest.log if you are crashing
- [X] Remove mod that enhances Minecraft: Optifine, Sodium, others. The issue persists.
- [X] If you are unsure which mod is the culprit.
Disable all of your mods and enable them 1-2 mods each time to isolate the culprit
- [X] Confirm that there is no existing issue with a similar description submitted in the list of issues.

### Version

1.20.1-FORGE

### Supplementaries Version

supplementaries-1.20-3.1.11

### Moonlight Lib Version

moonlight-1.20-2.13.44

### Describe Issue

I believe I narrowed down the issue I had earlier with long world load times. The higher the spawnradius gamemode is set at, the longer it takes to load into a world. I started with the default and ramped it up to my survival world's value of 50000 I was having issues with on the latest Supplementaries: 10, 100, 200, 300, 400, 500, 600, 700, 800, 900, 950, 999, 1000, 1500, 2000, 50000. Loading times were getting longer and longer until eventually I gave up waiting as I was already at 50 minutes of waiting to load on this minimal instance. Notice the time gap when `Changing view distance to 16, from 10` appears, and when I killed the instance in this log:
Latest Supplementaries Log: https://mclo.gs/ia0qoaD

Here on 2.8.17 and on a Vanilla instance with 50000 spawnradius set I load into the world almost instantly:
2.8.17 Supplementaries Log: https://mclo.gs/eyVehjz
Vanilla Instance Log: https://mclo.gs/AvIvyiE

It's not a big deal as I was setting a high spawnradius for a hardcore random spawn location. Was a more elegant solution than using a datapack that accomplished the same thing but in an ugly uncustomizable way. I'm just curious of what changed between version 2 and 3 of Supplementaries that causes load times to shoot up exponentially relative to the spawnradius (I tested every version above 2 and got the same issue).

**Comment by MehVahdJukaar** [comment]
Mind getting a spark report? One that starts when server is loading. Idk how to do that tbh but I know it's possible as I was sent those

**Comment by dandy-is-lion** [comment]
Sure. I'll see if I can get one with the default spawnradius and with a higher one that has a noticeable load time difference (that is hopefully less than 50 minutes).

**Comment by dandy-is-lion** [comment]
Here you go: https://spark.lucko.me/tg4Fg9jbRa

I started the profiler when I loaded into a normal world with default gamerules. Then set the spawnradius to 1000, logged back in and out. And one more time setting it to 5000. 1000 taking about 12 seconds to load, 5000 taking about 5 minutes. I stopped the profiler when I loaded in. Here's the log of that session: https://mclo.gs/ms8wojc

**Comment by dandy-is-lion** [comment]
Spark: https://spark.lucko.me/B8zZHQaqbQ
Ran another profiler with 5000 spawnradius which took about 5 minutes again, except I left it running for longer in case the profiler missed something. When I did finally log in, the world was taking a long time to load new chunks and I had trouble getting the profiler to stop. I'm aware that spark can cause high server load, and it was timing out a lot in the log: https://mclo.gs/llJ2T4W

**Comment by dandy-is-lion** [comment]
Tested again with 5000 spawnradius, but without spark running, and had the same issue of it taking 5 minutes to load in and new chunks taking a very long time to generate. The log for that: https://mclo.gs/GT0RiQh

**Comment by MehVahdJukaar** [comment]
i know whats causing this. quite peculiar as this particular use case proved to be quite effective at finding a quite hart to catch issue

**Comment by MehVahdJukaar** [comment]
use new moonlight lib soon

**Comment by dandy-is-lion** [comment]
Sweet. I'm curious to what the cause was too. I know my case is rare so I appreciate you troubleshooting to figure it out. Cheers and Happy New Year!

**Comment by MehVahdJukaar** [comment]
some fake level object generating more fake chunks than it needed to be. normally not an issue as for its use case at most 1 would be used. however its initialization would not as it would create all the ones that are requested by the spawn radius which is done on object creation.
Can you verify this version works for you?

**Comment by dandy-is-lion** [comment]
The latest MoonlightLib (2.13.45) appears to have partially resolved the issue. The load times are significantly faster, though they're not near-instantaneous like before. When initially generating a new world with spawnradius 50000, which is the worst case for me, I spawned in a void world with a huge TPS lag spike. After a minute or two however, the world loaded in like normal, and chunks were generating like normal, so that problem is partially fixed. Logging out and back in again still took a long time, about 2 minutes, but that's much faster than hour or so I would've waited before with this large of a spawnradius. However, this used to be near-instantaneous as well

The TPS lag spike upon world load after gen with 50000 spawnradius, initially I spawned in with nothing around me:
![2024-12-30_19 16 19](https://github.com/user-attachments/assets/8402bb6c-60fb-4235-b89e-21ffd8897d9a)

Log of me creating a world and then logging out and back in again: https://mclo.gs/7huh9MM

**Comment by MehVahdJukaar** [comment]
does this still happen?

**Comment by dandy-is-lion** [comment]
I'll check if this is still an issue when I have the chance later today

**Comment by dandy-is-lion** [comment]
Can confirm that on the latest versions of Supplementaries (3.1.18) and Moonlight (2.13.69) on Forge 1.20.1 the issue still persists. Tested on a minimal instance with these two mods and default worldgen, with the gamerule spawnRadius set to 50000. The initial world generation is very slow and spawns you in a void until after about 5 minutes you're dropped into a default world. Relogging had a similarly slow spawn in. However new chunks are generated like normal. Reducing or resetting the spawnRadius to default fixes both of these issues. Interestingly a high spawnRadius also causes TPS lag with Create's Deployers so I think the gamerule itself is a bit buggy, even if <3.0 versions of Supplementaries loaded worlds just fine with a high spawnRadius.

![Image](https://github.com/user-attachments/assets/f8188a82-c3d5-4228-af96-9c8814f7649f)

**Comment by MehVahdJukaar** [comment]
Mind sending the new spark report so I can see what's taking time? I hate fighting concurrency

**Comment by dandy-is-lion** [comment]
Sure thing. Here you go: https://spark.lucko.me/UZFumdltPz and the log just incase: https://mclo.gs/Z6mS4Kv

Interestingly I didn't spawn in a void world this time, but the initial worldgen and relog took just as long as last test. I relogged twice after the initial worldgen just to be sure. Same default worldgen with spawnRadius set to 50000.

![Image](https://github.com/user-attachments/assets/af662b19-8070-4520-99a5-75a4c15566f0)

**Comment by dandy-is-lion** [comment]
Let it run for a bit longer to see if caught anything else: https://spark.lucko.me/qWDL62Lj6h
It seems it has something to do with Fake Players and Overworld Respawning which makes sense with such a large spawnRadius. And that might explain why Create's Deployers are having issues too with that gamerule. I thought setting my spawn would help reduce the load times but it seems that it's on the Fake Player.

**Comment by MehVahdJukaar** [comment]
so thing is theres no more overhead compared to vailla impl. asyou can see here both take quite some time to place their player when big spawn radius is enabled as they try so many spawn positions. This means that evey time a player dies you would experience similar lag spike

![Image](https://github.com/user-attachments/assets/0621d567-054e-4b2b-af72-b189bf476fce)

That said i will add some code that will bypass thatpointles logic with all fake players, that would affect other mods that use them aswell. Tbh idk why forge isnt doing that already if that method is so expensive

**Comment by dandy-is-lion** [comment]
Ahh interesting! It probably hasn't been caught because a spawnRadius this high is so rare to use that it hasn't been reported to Forge yet. I wonder if the same thing occurs on Fabric?

I do know that in a completely vanilla instance, the worldgen is slightly longer than normal, but it's not this long. I'm assuming it's because of how thread expensive that process is for Fake Players

**Comment by MehVahdJukaar** [comment]
to add onto this its taking so long as it finds empty chunks and so it keeps on trying for entire spawn radius

**Comment by MehVahdJukaar** [comment]
oh and good call, deployers as you might imagine use fake players too. Thats how they simulate clicking. This thing im pushing should fix that too

**Comment by MehVahdJukaar** [comment]
is this solved?

**Comment by dandalous** [comment]
Yes it appears to be resolved. Including the issue with the Deployers at far distances. Thank you!
